rosidl_generator_cpp/cmake/rosidl_generator_cpp_generate_interfaces.cmake

```diff
@@ -74,20 +78,34 @@ rosidl_write_generator_arguments(
   TARGET_DEPENDENCIES ${target_dependencies}
 )
 
+set(generator_module_paths "")
+foreach (rosidl_generator_cpp_GENERATOR_FILE ${rosidl_generator_cpp_GENERATOR_FILES})
+  get_filename_component(generator_module_path ${rosidl_generator_cpp_GENERATOR_FILE} DIRECTORY)
+  get_filename_component(generator_module_path ${generator_module_path} DIRECTORY)
+  set(generator_module_paths "${generator_module_paths}:${generator_module_path}")
+endforeach()
+
+set(TMP_PYTHONPATH "$ENV{PYTHONPATH}${generator_module_paths}")
```

Comment:
I would suggest to not fiddle with the PYTHONPATH especially since these paths are not correct when `generator files` contains multiple files from different locations.

Instead you can update the actually invoked Python script to work from source. Instead of `from rosidl_generator_cpp import generate_cpp` you can do the following:

    try:
        from rosidl_generator_cpp import generate_cpp
    except ImportError:
        rosidl_generator_cpp_root = os.path.dirname(os.path.dirname(__file__)))
        rosidl_generator_cpp_package = os.path.join(rosidl_generator_cpp_root, 'rosidl_generator_cpp')
        if not os.path.exists(rosidl_generator_cpp_package):
            raise
        sys.path.insert(0, os.path.abspath(rosidl_generator_cpp_root))
        from rosidl_generator_cpp import generate_cpp